File: lib/not_wordle.py
import random
import sys
import json
import os

def load_words():
    """Load words from the words.txt file"""
    try:
        # Get the absolute path to the words.txt file
        words_file_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'public', 'files', 'words.txt')
        print(f"Loading words from: {words_file_path}", file=sys.stderr)
        
        with open(words_file_path, 'r') as word_file:
            words = [line.rstrip().upper() for line in word_file]
            print(f"Loaded {len(words)} words", file=sys.stderr)
            return words
    except FileNotFoundError:
        print(f"Words file not found at: {words_file_path}", file=sys.stderr)
        # Fallback to a small list of words if file not found
        return ["ABOUT", "ABOVE", "ABUSE", "ACTOR", "ACUTE", "ADMIT", "ADOPT", "ADULT", "AFTER", "AGAIN"]
    except Exception as e:
        print(f"Error loading words: {str(e)}", file=sys.stderr)
        # Fallback to a small list of words if there's an error
        return ["ABOUT", "ABOVE", "ABUSE", "ACTOR", "ACUTE", "ADMIT", "ADOPT", "ADULT", "AFTER", "AGAIN"]

def initialize_game():
    """Initialize a new game and return the word to guess"""
    word_list = load_words()
    word_to_guess = random.choice(word_list)
    print(f"Selected word to guess: {word_to_guess}", file=sys.stderr)
    return word_to_guess

def check_guess(word_to_guess, word_guessed):
    """Check a guess against the word to guess and return the result"""
    if word_guessed == word_to_guess:
        return {
            "correct": True,
            "message": f"Congratulations, you guessed the right word: {word_to_guess}",
            "correct_letters": list(word_to_guess),
            "out_of_place_letters": []
        }
    
    if len(word_guessed) != 5:
        return {
            "correct": False,
            "message": "Oops! Try again with a 5 letter word.",
            "correct_letters": [],
            "out_of_place_letters": []
        }
    
    letters_to_guess = list(word_to_guess)
    letters_guessed = list(word_guessed)
    correct_letters_guessed = ['-', '-', '-', '-', '-']
    out_of_place_letters_guessed = set()
    
    # First pass: find correct letters in correct positions
    for i in range(5):
        if letters_guessed[i] == letters_to_guess[i]:
            correct_letters_guessed[i] = letters_guessed[i]
            letters_to_guess[i] = '-'  # Mark as used
    
    # Second pass: find letters in wrong positions
    for i in range(5):
        if correct_letters_guessed[i] == '-':
            try:
                index_on_word_to_guess = letters_to_guess.index(letters_guessed[i])
                if index_on_word_to_guess >= 0:
                    out_of_place_letters_guessed.add(letters_guessed[i])
                    letters_to_guess[index_on_word_to_guess] = '-'  # Mark as used
            except ValueError:
                pass
    
    return {
        "correct": False,
        "message": "Keep guessing!",
        "correct_letters": correct_letters_guessed,
        "out_of_place_letters": list(out_of_place_letters_guessed)
    }

def main():
    """Main function to run the game in terminal mode"""
    word_to_guess = initialize_game()
    print('Welcome to "Definitely Not Just a Rip-Off of Wordle"!\nYou have 5 guesses to guess a 5 letter word.\nGood Luck!')
    
    correct_letters_guessed = ['-','-','-','-','-']
    out_of_place_letters_guessed = set()
    
    for turn in range(5):
        word_guessed = input('Please guess a word: ').upper()
        print('Your guess: {}'.format(word_guessed))
        
        result = check_guess(word_to_guess, word_guessed)
        
        if result["correct"]:
            print(result["message"])
            break
        else:
            print(result["message"])
            print(result["correct_letters"])
            if result["out_of_place_letters"]:
                print(result["out_of_place_letters"])
            
            if turn == 4:
                print(f"Game over! The word was: {word_to_guess}")

if __name__ == "__main__":
    main()

File: lib/games/not-wordle.ts
import fs from 'fs';
import path from 'path';

// Types
export interface GameState {
  wordToGuess: string;
  turns: number;
  maxTurns: number;
  gameOver: boolean;
}

export interface GuessResult {
  correct: boolean;
  message: string;
  correctLetters: string[];
  outOfPlaceLetters: string[];
}

// Store game state in memory (in a real app, you'd use a database)
const games: Record<string, GameState> = {};

/**
 * Load words from the words.txt file
 */
export function loadWords(): string[] {
  try {
    // Get the absolute path to the words.txt file
    const wordsFilePath = path.join(process.cwd(), 'public', 'files', 'words.txt');
    console.log(`Loading words from: ${wordsFilePath}`);
    
    const wordFile = fs.readFileSync(wordsFilePath, 'utf8');
    const words = wordFile.split('\n').map(line => line.trim().toUpperCase()).filter(word => word.length === 5);
    console.log(`Loaded ${words.length} words`);
    return words;
  } catch (error) {
    console.error(`Error loading words: ${error}`);
    // Fallback to a small list of words if there's an error
    return ["ABOUT", "ABOVE", "ABUSE", "ACTOR", "ACUTE", "ADMIT", "ADOPT", "ADULT", "AFTER", "AGAIN"];
  }
}

/**
 * Initialize a new game and return the word to guess
 */
export function initializeGame(): string {
  const wordList = loadWords();
  const wordToGuess = wordList[Math.floor(Math.random() * wordList.length)];
  console.log(`Selected word to guess: ${wordToGuess}`);
  return wordToGuess;
}

/**
 * Check a guess against the word to guess and return the result
 */
export function checkGuess(wordToGuess: string, wordGuessed: string): GuessResult {
  if (wordGuessed === wordToGuess) {
    return {
      correct: true,
      message: `Congratulations, you guessed the right word: ${wordToGuess}`,
      correctLetters: wordToGuess.split(''),
      outOfPlaceLetters: []
    };
  }
  
  if (wordGuessed.length !== 5) {
    return {
      correct: false,
      message: "Oops! Try again with a 5 letter word.",
      correctLetters: [],
      outOfPlaceLetters: []
    };
  }
  
  const lettersToGuess = wordToGuess.split('');
  const lettersGuessed = wordGuessed.split('');
  const correctLettersGuessed = ['-', '-', '-', '-', '-'];
  const outOfPlaceLettersGuessed = new Set<string>();
  
  // First pass: find correct letters in correct positions
  for (let i = 0; i < 5; i++) {
    if (lettersGuessed[i] === lettersToGuess[i]) {
      correctLettersGuessed[i] = lettersGuessed[i];
      lettersToGuess[i] = '-';  // Mark as used
    }
  }
  
  // Second pass: find letters in wrong positions
  for (let i = 0; i < 5; i++) {
    if (correctLettersGuessed[i] === '-') {
      const indexOnWordToGuess = lettersToGuess.indexOf(lettersGuessed[i]);
      if (indexOnWordToGuess >= 0) {
        outOfPlaceLettersGuessed.add(lettersGuessed[i]);
        lettersToGuess[indexOnWordToGuess] = '-';  // Mark as used
      }
    }
  }
  
  return {
    correct: false,
    message: "Keep guessing!",
    correctLetters: correctLettersGuessed,
    outOfPlaceLetters: Array.from(outOfPlaceLettersGuessed)
  };
}

/**
 * Handle a new game request
 */
export function handleNewGame(gameId: string): { status: string; message: string; gameId?: string } {
  if (!gameId) {
    console.error("Error: No game_id provided for new_game action");
    return {
      status: 'error',
      message: 'Game ID is required for new_game action'
    };
  }
  
  const wordToGuess = initializeGame();
  games[gameId] = {
    wordToGuess,
    turns: 0,
    maxTurns: 5,
    gameOver: false
  };
  
  console.log(`Created new game with ID: ${gameId}, word: ${wordToGuess}`);
  console.log(`Games after new_game: ${JSON.stringify(games)}`);
  
  return {
    status: 'success',
    message: '',
    gameId
  };
}

/**
 * Handle a guess request
 */
export function handleGuess(gameId: string, guess: string): {
  status: string;
  message: string;
  correct?: boolean;
  correctLetters?: string[];
  outOfPlaceLetters?: string[];
  turnsLeft?: number;
  word?: string;
} {
  if (!gameId) {
    console.error("Error: No game_id provided for guess action");
    return {
      status: 'error',
      message: 'Game ID is required for guess action'
    };
  }
  
  const wordGuessed = guess.toUpperCase();
  
  if (!games[gameId]) {
    console.error(`Game not found with ID: ${gameId}`);
    console.error(`Available games: ${Object.keys(games)}`);
    console.error(`Games state: ${JSON.stringify(games)}`);
    return {
      status: 'error',
      message: `Game not found with ID: ${gameId}. Start a new game first.`
    };
  }
  
  const game = games[gameId];
  console.log(`Processing guess for game ${gameId}: ${wordGuessed}`);
  console.log(`Game state: ${JSON.stringify(game)}`);
  
  if (game.gameOver) {
    return {
      status: 'error',
      message: 'Game is already over. Start a new game.'
    };
  }
  
  if (game.turns >= game.maxTurns) {
    game.gameOver = true;
    // Get the result for the final guess to show correct/incorrect letters
    const result = checkGuess(game.wordToGuess, wordGuessed);
    return {
      status: 'game_over',
      message: `Game over! The word was: ${game.wordToGuess}`,
      word: game.wordToGuess,
      correctLetters: result.correctLetters,
      outOfPlaceLetters: result.outOfPlaceLetters
    };
  }
  
  const result = checkGuess(game.wordToGuess, wordGuessed);
  game.turns += 1;
  
  if (result.correct) {
    game.gameOver = true;
    console.log(`Player won game ${gameId} in ${game.turns} turns`);
    return {
      status: 'success',
      message: result.message,
      correct: true,
      correctLetters: result.correctLetters,
      outOfPlaceLetters: result.outOfPlaceLetters,
      turnsLeft: game.maxTurns - game.turns
    };
  }
  
  if (game.turns >= game.maxTurns) {
    game.gameOver = true;
    console.log(`Player lost game ${gameId}`);
    // Get the result for the final guess to show correct/incorrect letters
    const result = checkGuess(game.wordToGuess, wordGuessed);
    return {
      status: 'game_over',
      message: `Game over! The word was: ${game.wordToGuess}`,
      word: game.wordToGuess,
      correctLetters: result.correctLetters,
      outOfPlaceLetters: result.outOfPlaceLetters
    };
  }
  
  console.log(`Guess processed for game ${gameId}, turns left: ${game.maxTurns - game.turns}`);
  return {
    status: 'success',
    message: result.message,
    correct: false,
    correctLetters: result.correctLetters,
    outOfPlaceLetters: result.outOfPlaceLetters,
    turnsLeft: game.maxTurns - game.turns
  };
}

/**
 * Handle a request from the Next.js app
 */
export function handleRequest(requestData: { action: string; game_id?: string; guess?: string }): {
  status: string;
  message: string;
  gameId?: string;
  correct?: boolean;
  correctLetters?: string[];
  outOfPlaceLetters?: string[];
  turnsLeft?: number;
  word?: string;
} {
  // Debug logging
  console.log(`Received request: ${JSON.stringify(requestData)}`);
  
  const { action, game_id, guess } = requestData;
  
  // Debug logging
  console.log(`Action: ${action}, Game ID: ${game_id}`);
  console.log(`Current games: ${Object.keys(games)}`);
  
  if (action === 'new_game') {
    return handleNewGame(game_id || '');
  } else if (action === 'guess') {
    return handleGuess(game_id || '', guess || '');
  } else {
    console.error(`Invalid action: ${action}`);
    return {
      status: 'error',
      message: 'Invalid action'
    };
  }
} 